src/pattern-library/components/patterns/data/DataTablePage.tsx

```diff
@@ -1,17 +1,87 @@
-import { useState } from 'preact/hooks';
+import { useCallback, useMemo, useState } from 'preact/hooks';
 
-import { DataTable, Scroll } from '../../../../';
+import { DataTable, type DataTableProps, Scroll } from '../../../../';
+import type { Order } from '../../../../components/data/DataTable';
 import Library from '../../Library';
 import { nabokovNovels } from '../samples';
 import type { NabokovNovel } from '../samples';
 
 const nabokovRows = nabokovNovels();
 const nabokovColumns = [
-  { field: 'title', label: 'Title' },
-  { field: 'year', label: 'Year' },
-  { field: 'language', label: 'Language' },
+  { field: 'title' as const, label: 'Title' },
+  { field: 'year' as const, label: 'Year' },
+  { field: 'language' as const, label: 'Language' },
 ];
 
+type StrictNabokovNovel = Omit<NabokovNovel, 'translatedTitle'>;
+
+function useOrderedRows(
+  rows: StrictNabokovNovel[],
+  order: Order<keyof StrictNabokovNovel>,
+) {
+  return useMemo(() => {
+    if (!order) {
+      return rows;
+    }
+
+    return [...rows].sort((a, b) => {
+      if (a[order.field] === b[order.field]) {
+        return 0;
+      }
+
+      if (order.direction === 'asc') {
+        return a[order.field] > b[order.field] ? 1 : -1;
+      }
+
+      return a[order.field] > b[order.field] ? -1 : 1;
+    });
+  }, [order, rows]);
+}
+
+function ClientOrderedDataTable({
+  rows,
+  ...rest
+}: Omit<DataTableProps<StrictNabokovNovel>, 'order' | 'onOrderChange'>) {
+  const [order, setOrder] = useState<Order<keyof StrictNabokovNovel>>(null);
+  const orderedRows = useOrderedRows(rows, order);
+
+  return (
+    <DataTable
+      {...rest}
+      rows={orderedRows}
+      order={order}
+      onOrderChange={setOrder}
+    />
+  );
+}
+
+function AsyncOrderableDataTable({
+  rows,
+  ...rest
+}: Omit<DataTableProps<StrictNabokovNovel>, 'order' | 'onOrderChange'>) {
+  const [loading, setLoading] = useState(false);
+  const [order, setOrder] = useState<Order<keyof StrictNabokovNovel>>(null);
+  const changeOrder = useCallback(
+    (newOrder: Order<keyof StrictNabokovNovel>) => {
+      setOrder(newOrder);
+      setLoading(true);
+      setTimeout(() => setLoading(false), 600);
```

Comment:
The async loading of ordered rows is faked here.

In a real scenario, we would be using a fetcher, where the order would be part of the key somehow, and would also receive the order object. The we would pass the list of rows returned by the fetcher asynchronously, and use its `loading` to provide it to the DataTable.

Reply:
If you used the same approach as the fake implementation here but used varying timeouts then you might get the wrong result if clicking to change the order while the table is loading. If the order becomes part of the key in a proper fetch mechanism then that problem should be avoided.

Reply:
Good catch. I'll try to handle that here.

I guess in the case of using `useFetch`, the in-progress request would be aborted and a new one would be triggered, so we should be covered as you said.